src/JsonSchema/SchemaFactory.php

```diff
@@ -214,7 +214,7 @@ private function buildPropertySchema(Schema $schema, string $definitionName, str
                 $className = $valueType->getClassName();
             }
 
-            $valueSchema = $this->typeFactory->getType(new Type($builtinType, $type->isNullable(), $className, $isCollection), $format, $propertyMetadata->isReadableLink(), $serializerContext, $schema);
+            $valueSchema = $this->typeFactory->getType(new Type($builtinType, !$propertyMetadata->isRequired() || $type->isNullable(), $className, $isCollection), $format, $propertyMetadata->isReadableLink(), $serializerContext, $schema);
```

Comment:
Being nullable is not enought here, we also need to check if the property is required.
If it's not required, it means this property should be nullable.